### PR TITLE
Laboratory Work #1, Nadezhda Burakova - 20FPL2

### PR DESCRIPTION
**Pull request by Burakovan**

### CODE REVIEW

File: pipeline.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 """
 Pipeline for text processing implementation
```
Comment (marina-kaz) [inline pipeline.py:2]: You are working with a wrong file
You are supposed to implement scrapper logic during current part of laboratory work

**Review by marina-kaz — CHANGES_REQUESTED** [review]
There is not much time left so I advise to speed up a little :)

File: requirements.txt [hunk 2]
```diff
@@ -1 +1,3 @@
-
+beautifulsoup4==4.10.0
+lxml
+requests
```
Comment (marina-kaz) [inline requirements.txt:3]: it is better to specify versions as well

File: scrapper.py [hunk 3]
```diff
@@ -1,6 +1,14 @@
...
 Scrapper implementation
 """
+import json
+import re
+import shutil
+import requests
+from pathlib import Path
+from bs4 import BeautifulSoup
+from constants import ASSETS_PATH, CRAWLER_CONFIG_PATH
+from core_utils.article import Article
```
Comment (marina-kaz) [inline scrapper.py:12]: it is a good practice to sort imports alphabetically

File: scrapper.py [hunk 4]
```diff
@@ -25,39 +33,108 @@ class Crawler:
...
+        for seed_url in self.seed_urls:
+            response = requests.get(url=seed_url)
+
+            if not response.ok:
+                continue
+
+            soup = BeautifulSoup(response.text, 'lxml')
+            article_urls = self._extract_url(soup)
+            for article_url in article_urls:
+                self.urls.append(article_url)
```
Comment (marina-kaz) [inline scrapper.py:65]: check that you do not save duplicating links

Comment (marina-kaz) [reply scrapper.py:65]: also, what if you reach the required number of links before you end this cycle?

File: scrapper.py [hunk 5]
```diff
@@ -25,39 +33,108 @@ class Crawler:
...
+        self.article.text = ''
+        for p in paragraphs:
+            self.article.text += p.text
+
+    def parse(self):
+        response = requests.get(url=self.article_url)
+        article_bs = BeautifulSoup(response.text, 'lxml')
+
+        self._fill_article_with_text(article_bs)
+        self.article.save_raw()
```
Comment (marina-kaz) [inline scrapper.py:92]: Parser is not responsible for saving articles. It must parse and return only.

File: scrapper.py [hunk 6]
```diff
@@ -25,39 +33,108 @@ class Crawler:
...
     """
-    pass
+    with open(crawler_path) as file:
+        config = json.load(file)
+    max_articles = config["total_articles_to_find_and_parse"]
+    seed_urls = config["seed_urls"]
+
+    if not seed_urls:
+        raise IncorrectURLError
+    for article_url in seed_urls:
```
Comment (marina-kaz) [inline scrapper.py:118]: what if seed_urls is not an iterable? process this case, too

**Review by marina-kaz — CHANGES_REQUESTED** [review]
Not bad! Keep going

**Review by marina-kaz — CHANGES_REQUESTED** [review]
A little polishing would help

**Review by marina-kaz — CHANGES_REQUESTED** [review]
Time to finish you Crawler and embark on Pipeline

File: scrapper.py (hunk 3 above)
Comment (marina-kaz) [reply scrapper.py:13]: according to alphabet, _b_ precedes _r_

File: scrapper.py [hunk 7]
```diff
@@ -25,39 +34,133 @@ class Crawler:
...
 
 if __name__ == '__main__':
     # YOUR CODE HERE
-    pass
+    seed_links, maximum_articles = validate_config(CRAWLER_CONFIG_PATH)
+    prepare_environment(ASSETS_PATH)
+    crawler = Crawler(seed_urls=seed_links, max_articles=maximum_articles)
+    crawler.find_articles()
+
+    COUNTER = 1
```
Comment (marina-kaz) [inline scrapper.py:161]: in python, we usually use capital letters to denote constants 
you are changing the value of `COUNTER ` after each iteration, so it is not exactly constant 
I would recommend to use `enumerate()` for declaring a variable that would serve as an ascending index

**Review by marina-kaz — CHANGES_REQUESTED** [review]
Good job!

**Review by marina-kaz — CHANGES_REQUESTED** [review]
Time to implement Pipeline as well

**Review by marina-kaz — CHANGES_REQUESTED** [review]
Nothing to review yet

File: pipeline.py [hunk 8]
```diff
@@ -72,32 +87,110 @@ class TextProcessingPipeline:
...
+            multiple_tagged_tokens = []
+
+            for token in tokens:
+                tokens_for_article.append(token.get_cleaned())
+                single_tagged_tokens.append(token.get_single_tagged())
+                multiple_tagged_tokens.append(token.get_multiple_tagged())
+
+            article.save_as(' '.join(tokens_for_article), 'cleaned')
+            article.save_as(' '.join(single_tagged_tokens), 'single_tagged')
+            article.save_as(' '.join(multiple_tagged_tokens), 'multiple_tagged')
```
Comment (marina-kaz) [inline pipeline.py]: as it is stated in the 5.3 section of pipeline documentation, you are required to pass `ArtifactType` attributes as a `kind` argument to the `save_as() method`, this way code is more secure as it is protected from almost untraceable errors caused by simple typos

File: pipeline.py [hunk 9]
```diff
@@ -72,32 +87,110 @@ class TextProcessingPipeline:
...
+    all_article_ids = []
+
+    for file in path.glob("*.txt"):
+        with open(file, 'r', encoding='utf-8') as text_file:
+            text = text_file.read()
+        raw_data = list(path.glob('*_raw.txt'))
+        meta_data = list(path.glob('*_meta.json'))
+
+        if not len(meta_data) == len(raw_data):
+            raise InconsistentDatasetError
```
Comment (marina-kaz) [inline pipeline.py]: could you please explain why you have to verify this condition within a cycle over files

**Review by marina-kaz — CHANGES_REQUESTED** [review]
Good job, yet it looks unusually similar to another student's work

File: pipeline.py (hunk 9 above)
Comment (Burakovan) [reply pipeline.py]: To make sure that i got raw texts from all of the files that i have

File: pipeline.py (hunk 8 above)
Comment (Burakovan) [reply pipeline.py]: I fixed it

**Pull request closed without merging**